Crashlytics/Crashlytics/Models/Record/FIRCLSReportAdapter.m

```diff
@@ -388,13 +448,66 @@ - (google_crashlytics_Session_Event_Application)protoEventApplication {
   app.background = [self wasInBackground];
   app.ui_orientation = [self uiOrientation];
 
-  // TODO: Where do these come from?
-  //    app.custom_attributes
-  //    app.custom_attributes_count
+  app.custom_attributes = [self protoCustomAttributesWithError:nil];
+  app.custom_attributes_count = (pb_size_t)self.userKeyValues.count;
 
   return app;
 }
 
+- (google_crashlytics_Session_Event_Application)protoEventApplicationForError:
+    (FIRCLSRecordError *)error {
+  google_crashlytics_Session_Event_Application app =
+      google_crashlytics_Session_Event_Application_init_default;
+
+  // TODO: Filter by binaries by stacktrace, Protobuf.scala:93
+  app.execution.binaries = [self protoBinaryImages];
+  app.execution.binaries_count = (pb_size_t)self.binaryImages.count;
+
+  google_crashlytics_Session_Event_Application_Execution_Signal emptySignal =
+      google_crashlytics_Session_Event_Application_Execution_Signal_init_default;
+  app.execution.signal = emptySignal;
+
+  // Create single thread from stacktrace
+  google_crashlytics_Session_Event_Application_Execution_Thread *threads =
+      malloc(sizeof(google_crashlytics_Session_Event_Application_Execution_Thread) * 1);
+  google_crashlytics_Session_Event_Application_Execution_Thread thread =
+      google_crashlytics_Session_Event_Application_Execution_Thread_init_default;
+  thread.frames = [self protoFramesWithStacktrace:error.stacktrace];
+  thread.frames_count = (pb_size_t)error.stacktrace.count;
+  threads[0] = thread;
+  app.execution.threads = threads;
+  app.execution.threads_count = 1;
+
+  app.background = [self wasInBackground];
+  app.ui_orientation = [self uiOrientation];
+
+  app.custom_attributes = [self protoCustomAttributesWithError:error];
+  app.custom_attributes_count = (pb_size_t)self.userKeyValues.count + 2;
```

Comment:
Here, line 452, and line 495 need to be reconciled - I'm worried we are gonna get our counts wrong with what's allocated in the arrays.
 - If we change the number inside `protoCustomAttributesWithError`, the +2 on this line may get out of sync
 - As a reader, I'm unsure the relationship between `self.userKeyValues.count` and `kvs.count`on 495.

A couple suggestions is to add an argument `* count` to `protoCustomAttributesWithError` and write to the value inside the method, or pass the count into `protoCustomAttributesWithError` as an argument, or multiple returns?

Reply:
I now pass the key values into protoCustomAttributes